modules/channel-web/src/migrations/v12_24_0-1621284696-messaging.ts

```diff
@@ -0,0 +1,758 @@
+    await this.createClients()
+  }
+
+  async cleanup() {
+    await this.bp.database.schema.dropTable('web_messages')
+    await this.bp.database.schema.dropTable('web_conversations')
+  }
+
+  protected async createTables() {
+    // We need to create the messaging tables here because the messaging
+    // server isn't started before we run the migrations
+
+    await this.bp.database.schema.createTable('msg_providers', table => {
+      table.uuid('id').primary()
+      table
+        .string('name')
+        .unique()
+        .notNullable()
+      table.boolean('sandbox').notNullable()
+    })
+
+    await this.bp.database.schema.createTable('msg_clients', table => {
+      table.uuid('id').primary()
+      table
+        .uuid('providerId')
+        .references('id')
+        .inTable('msg_providers')
+        .unique()
+        .notNullable()
+      table
+        .string('token')
+        .unique()
+        .notNullable()
+    })
+
+    await this.bp.database.schema.createTable('msg_users', table => {
+      table.uuid('id').primary()
+      table
+        .uuid('clientId')
+        .references('id')
+        .inTable('msg_clients')
+        .notNullable()
+    })
+
+    await this.bp.database.schema.createTable('msg_conversations', table => {
+      table.uuid('id').primary()
+      table
+        .uuid('clientId')
+        .references('id')
+        .inTable('msg_clients')
+        .notNullable()
+      table
+        .uuid('userId')
+        .references('id')
+        .inTable('msg_users')
+        .notNullable()
+      table.timestamp('createdOn').notNullable()
+      table.index(['userId', 'clientId'])
+    })
+
+    await this.bp.database.schema.createTable('msg_messages', table => {
+      table.uuid('id').primary()
+      table
+        .uuid('conversationId')
+        .references('id')
+        .inTable('msg_conversations')
+        .notNullable()
+        .onDelete('cascade')
+      table
+        .uuid('authorId')
+        .references('id')
+        .inTable('msg_users')
+        .nullable()
+      table.timestamp('sentOn').notNullable()
+      table.jsonb('payload').notNullable()
+      table.index(['conversationId', 'sentOn'])
+    })
+
+    // We need to create this here because sometimes the migration is ran before the module is initalized
+    await this.bp.database.schema.createTable('web_user_map', table => {
+      table.string('visitorId').primary()
+      table.uuid('userId').unique()
+    })
+  }
+
+  protected async createClients() {
+    const bots = await this.bp.bots.getAllBots()
+
+    for (const bot of bots.values()) {
+      const provider = {
+        id: uuid.v4(),
+        name: bot.id,
+        sandbox: false
+      }
+
+      await this.bp.database('msg_providers').insert(provider)
+
+      const token = crypto.randomBytes(66).toString('base64')
+      const client = {
+        id: uuid.v4(),
+        providerId: provider.id,
+        token: await bcrypt.hash(token, 10)
+      }
```

Comment:
What's the difference between a provider and a client, since they share the botId ? I'd really like to have a small definition / database schema to understand the whole setup. Couldn't the token be put in provider instead?